### PR TITLE
Merge duplicate key fix

### PR DESCRIPTION
**Pull request by acquamarin**
Fixes merge duplicate primary key error issue #3103. 
We now use aggregate hash table to indicate the duplicated primary keys. If a key is mark as duplicated, we skip inserting it.

### CODE REVIEW

File: src/processor/map/map_aggregate.cpp [hunk 1]
```diff
@@ -93,28 +94,80 @@ std::unique_ptr<PhysicalOperator> PlanMapper::mapAggregate(LogicalOperator* logi
     }
 }
 
+std::unique_ptr<PhysicalOperator> PlanMapper::mapMarkAccumulate(LogicalOperator* op) {
```
Comment (andyfengHKU) [inline src/processor/map/map_aggregate.cpp:97]: let's put this into its own file maybe?

File: src/processor/operator/aggregate/aggregate_hash_table.cpp [hunk 2]
```diff
@@ -212,6 +202,140 @@ void AggregateHashTable::resize(uint64_t newSize) {
...
+                        vector, idx, hashSlotsToUpdateAggState[idx]->entry + colOffset)) {
+                    mayMatchIdxes[mayMatchIdx++] = idx;
+                } else {
+                    noMatchIdxes[numNoMatches++] = idx;
+                }
+            }
+        } else {
+            for (auto i = 0u; i < numMayMatches; i++) {
+                auto idx = mayMatchIdxes[i];
+                auto isEntryKeyNull = factorizedTable->isNonOverflowColNull(
```
Comment (andyfengHKU) [inline src/processor/operator/aggregate/aggregate_hash_table.cpp:266]: Can u have a separate PR to cover the uncovered lines.

File: src/processor/operator/persistent/merge.cpp [hunk 3]
```diff
@@ -29,15 +32,30 @@ bool Merge::getNextTuplesInternal(ExecutionContext* context) {
...
+    KU_ASSERT(existenceVector->state->isFlat());
+    auto existencePos = existenceVector->state->selVector->selectedPositions[0];
+    if (!existenceVector->isNull(existencePos)) {
         for (auto& executor : onMatchNodeSetExecutors) {
             executor->set(context);
         }
         for (auto& executor : onMatchRelSetExecutors) {
             executor->set(context);
         }
+    } else if (distinctVector != nullptr &&
```
Comment (andyfengHKU) [inline src/processor/operator/persistent/merge.cpp:44]: Better write as
```
else {  // pattern not exist
     if (distinctVector ...) { // pattern has been created
          // Do on match and write result
     } else {
           // Do insert and on create.
      }
}
```

Also let's cover the on match and on create case in test

**Review by andyfengHKU — APPROVED** [review]

File: src/planner/operator/logical_mark_accumulate.cpp [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+void LogicalMarkAccumulate::computeFactorizedSchema() {
+    createEmptySchema();
+    auto childSchema = children[0]->getSchema();
+    // TODO(Ziyi): this may cause a bug, I'm not sure.
+    SinkOperatorUtil::recomputeSchema(*childSchema, childSchema->getExpressionsInScope(), *schema);
+    f_group_pos groupPos;
+    if (!keys.empty()) {
+        groupPos = schema->getGroupPos(*keys[0]);
+    } else {
+        groupPos = schema->createGroup();
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/planner/operator/logical_mark_accumulate.cpp:20]: As we discussed, let's throw an exception for now

**Pull request merged**